### DISCUSSION TITLE
extract image type

### DISCUSSION THREAD

**Discussion by NathanTech7713** [jsvine/pdfplumber#514] [Q&A]
Hi there,

I was wondering if there is a way to get the image format from the pdf?

Take the below code for example:
>>> import pdfplumber
>>> pdf=pdfplumber.open("my_pdf.pdf")
image=pdf.images[0]

As it stands, you can currently do:
image_data=image["stream"].get_data()

But without knowing the type of that image, I don't see how you could save that to a separate file or display it?

relatedly, I'd love to be able to contribute to this image object as I think making it an object rather than a dictionary would make life so much easier.
For example instead of:
image["stream"].get_data()
simply have:
image.get_data()

I think I have the coding knowledge, but don't understand the contributing requirements that well.
thanks in advance.
Nathan

**Comment by jsvine** [comment]
Hi @NathanTech7713, and very interesting question — thanks for raising it! I don't spend much time working with images in PDFs, so I don't have great answers for this, but it's worth discussing/exploring.

> But without knowing the type of that image, I don't see how you could save that to a separate file or display it?

It looks like `pdfminer.six` does have methods for obtaining an image file extension — see https://github.com/pdfminer/pdfminer.six/blob/c8cceb7c58deec9e647be6d3957e03442770bdd0/pdfminer/image.py#L140-L154. It would probably be possible to write a `pdfplumber.utils` method to do the same, as we are [already extracting the necessary attributes](https://github.com/jsvine/pdfplumber/blob/002500ab2e28885af1fa6e1b13c5234f874bcdae/pdfplumber/page.py#L27-L44) (`bits`, `colorspace`, and `stream`)

> relatedly, I'd love to be able to contribute to this image object as I think making it an object rather than a dictionary would make life so much easier.

It's important, for the rest of `pdfplumber`, that all extracted page objects are represented as simple `dict`s — at least under the library's current architecture. Instead, if you'd like to add image-specific functionality, I'd recommend adding a `pdfplumber.utils` method. Riffing on your example above:

```python
def save_image(image, path):
  with open(path, "wb") as f:
    data = image["stream"].get_data()
    f.write(data)
```

> I think I have the coding knowledge, but don't understand the contributing requirements that well.

Please see https://github.com/jsvine/pdfplumber/blob/stable/CONTRIBUTING.md

If you have questions that are not answered there, please let me know and I can try to answer them.
